ledger/sandbox/src/main/scala/com/digitalasset/platform/apiserver/services/ApiSubmissionService.scala

```diff
@@ -99,10 +114,49 @@ final class ApiSubmissionService private (
     val failedInterpretationsMeter: Meter =
       metrics.meter("daml.lapi.command_submission_service.failed_command_interpretations")
 
+    val deduplicatedCommandsMeter: Meter =
+      metrics.meter("daml.lapi.command_submission_service.deduplicated_commands")
+
     val submittedTransactionsTimer: Timer =
       metrics.timer("daml.lapi.command_submission_service.submitted_transactions")
   }
 
+  private def deduplicateAndRecordOnLedger(commands: ApiCommands)(
+      implicit logCtx: LoggingContext): Future[Unit] = {
+    val deduplicationKey = commands.submitter + "%" + commands.commandId.unwrap
+    val submittedAt = Instant.now
```

Comment:
Would it make sense to pass in a clock in the constructor to more easily drive tests?

Reply:
The command deduplication window should be driven by UTC time. If we were to change that to a custom clock, we would need to make sure the ledger (which can do additional command deduplication) uses the same clock.